pkg/controllers/failuredomainrollout/controller.go

```diff
@@ -0,0 +1,364 @@
+	options *controller.Options,
+) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.Cluster{}).
+		Watches(
+			&controlplanev1.KubeadmControlPlane{},
+			handler.EnqueueRequestsFromMapFunc(r.kubeadmControlPlaneToCluster),
+			builder.WithPredicates(predicate.ResourceVersionChangedPredicate{}),
+		).
+		WithOptions(*options).
+		Complete(r)
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx).WithValues("cluster", req.NamespacedName)
+	logger.V(5).Info("Starting failure domain rollout reconciliation")
+
+	var cluster clusterv1.Cluster
+	if err := r.Get(ctx, req.NamespacedName, &cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(5).Info("Cluster not found, skipping reconciliation")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get Cluster %s: %w", req.NamespacedName, err)
+	}
+
+	// Early validation checks
+	if cluster.Spec.Topology == nil {
+		logger.V(5).Info("Cluster is not using topology, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	if cluster.Spec.ControlPlaneRef == nil {
+		logger.V(5).Info("Cluster has no control plane reference, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	if len(cluster.Status.FailureDomains) == 0 {
+		logger.V(5).Info("Cluster has no failure domains, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	// If the Cluster is not fully reconciled, we should skip our own reconciliation.
+	if cluster.Status.ObservedGeneration < cluster.Generation {
+		logger.V(5).Info("Cluster is not yet reconciled, skipping failure domain rollout check",
+			"observedGeneration", cluster.Status.ObservedGeneration, "generation", cluster.Generation)
+		return ctrl.Result{RequeueAfter: 2 * time.Minute}, nil
+	}
+
+	// Get the KubeAdmControlPlane
+	kcpKey := types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Spec.ControlPlaneRef.Name,
+	}
+
+	var kcp controlplanev1.KubeadmControlPlane
+	if err := r.Get(ctx, kcpKey, &kcp); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.V(5).Info("KubeAdmControlPlane not found, skipping reconciliation")
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, fmt.Errorf("failed to get KubeAdmControlPlane %s: %w", kcpKey, err)
+	}
+
+	// If the KubeadmControlPlane is not fully reconciled, we should skip our own reconciliation.
+	if kcp.Status.ObservedGeneration < kcp.Generation {
+		logger.V(5).Info("KubeAdmControlPlane is not yet reconciled, skipping failure domain rollout check",
+			"observedGeneration", kcp.Status.ObservedGeneration, "generation", kcp.Generation)
+		return ctrl.Result{RequeueAfter: 2 * time.Minute}, nil
+	}
+
+	// Check if we need to trigger a rollout
+	needsRollout, reason, err := r.shouldTriggerRollout(ctx, &cluster, &kcp)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to determine if rollout is needed: %w", err)
+	}
+
+	if needsRollout {
```

Comment:
nit: Invert this logic to return early.

```suggestion
	if !needsRollout {
        return ctrl.Result{}, nil
    }
```